llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -263,9 +254,70 @@ bool RepoPruning::runOnModule(Module &M) {
     }
   }
 
-  LLVM_DEBUG(dbgs() << "size of module: " << M.size() << '\n');
-  LLVM_DEBUG(dbgs() << "size of removed functions: " << NumFunctions << '\n');
-  LLVM_DEBUG(dbgs() << "size of removed variables: " << NumVariables << '\n');
+  LLVM_DEBUG(dbgs() << "Size of module: " << M.size() << '\n');
+  LLVM_DEBUG(dbgs() << "Number of removed functions: " << NumFunctions << '\n');
+  LLVM_DEBUG(dbgs() << "Number of removed variables: " << NumVariables << '\n');
 
   return Changed;
 }
+
+static bool wasPruned(const GlobalObject &GO) {
+  if (const MDNode *const MD = GO.getMetadata(LLVMContext::MD_repo_ticket)) {
+    if (const TicketNode *const TN = dyn_cast<TicketNode>(MD)) {
+      return TN->getPruned();
+    }
+  }
+  return false;
+}
+
+static bool eliminateUnreferencedAndPrunedGOs(Module &M) {
+  bool Changed = false;
+
+  // Drop the bodies of available externally functions.
+  for (Function &F : M) {
```

Comment:
This loop goes through all of the functions in-order. The doPruning() code also does that. Is there a reason why we first mark functions as available-extern then remove some of them. Could the loops be folded together? (Same applies to the variable loops, of course.) I ask partly because the `wasPruned()` below calls `GO.getMetadata` which does a search of the`GlobalObjectMetadata` dense-map. If the loops were folded, we may be able to eliminate this search.

Reply:
Good point. The patch has been updated.